test/bio_tfo_test.c

```diff
@@ -0,0 +1,398 @@
+        server_tfo = 1;
+        client_tfo = 1;
+        break;
+    case 3:
+        client_tfo = 1;
+        break;
+    case 4:
+        server_tfo = 1;
+        break;
+    }
+
+    /* ACCEPT SOCKET */
+    if (!TEST_ptr(abio = BIO_new_accept("localhost:0"))
+            || !TEST_true(BIO_set_nbio_accept(abio, 1))
+            || !TEST_true(BIO_set_tfo_accept(abio, server_tfo))
+            || !TEST_int_gt(BIO_do_accept(abio), 0)
+            || !TEST_ptr(port = BIO_get_accept_port(abio)))
+        goto err;
+
+    /* Note: first BIO_do_accept will basically do the bind/listen */
+
+    /* CLIENT SOCKET */
+    if (!TEST_ptr(cbio = BIO_new_connect("localhost"))
+            || !TEST_long_gt(BIO_set_conn_port(cbio, port), 0)
+            || !TEST_long_gt(BIO_set_nbio(cbio, 1), 0)
+            || !TEST_long_gt(BIO_set_tfo(cbio, client_tfo), 0))
+        goto err;
+
+    /* FIRST ACCEPT: no connection should be established */
+    if (BIO_do_accept(abio) <= 0) {
+        if (!BIO_should_retry(abio)) {
+            sockerr = get_last_socket_error();
+            BIO_printf(bio_err, "Error: failed without EAGAIN\n");
+            goto err;
+        }
+    } else {
+        sbio = BIO_pop(abio);
+        BIO_printf(bio_err, "Error: accepted unknown connection\n");
+        goto err;
+    }
+
+    /* CONNECT ATTEMPT: different behavior based on TFO support */
+    if (BIO_do_connect(cbio) <= 0) {
+        sockerr = get_last_socket_error();
+        if (sockerr == EOPNOTSUPP) {
+            BIO_printf(bio_err, "Skip: TFO not enabled/supported for client\n");
+            goto success;
+        } else if (sockerr != EINPROGRESS) {
+            BIO_printf(bio_err, "Error: failed without EINPROGRESSn");
+            goto err;
+        }
+    }
+
+    /* macOS needs some time for this to happen, so put in a select */
+    tmp = BIO_wait(abio, time(NULL)+2, 0);
+    sockerr = get_last_socket_error();
+    if (!TEST_int_ge(tmp, 0)) {
+        BIO_printf(bio_err, "Error: socket wait failed\n");
+        goto err;
+    }
+
+    /* SECOND ACCEPT: if TFO is supported, this will still fail until data is sent */
+    if (BIO_do_accept(abio) <= 0) {
+        if (!BIO_should_retry(abio)) {
+            sockerr = get_last_socket_error();
+            BIO_printf(bio_err, "Error: failed without EAGAIN\n");
+            goto err;
+        }
+    } else {
+        if (idx == 0)
+            BIO_printf(bio_err, "Success: non-TFO connection accepted without data\n");
+        else if (idx == 1)
+            BIO_printf(bio_err, "Ignore: connection accepted before data, possibly no TFO cookie, or TFO may not be enabled\n");
+        else
+            BIO_printf(bio_err, "Warning: connection accepted before data, TFP may not be enabled\n");
+        sbio = BIO_pop(abio);
+        goto success;
+    }
+
+    /* SEND DATA: this should establish the actual TFO connection */
+    if (!TEST_true(BIO_write_ex(cbio, SOCKET_DATA, SOCKET_DATA_LEN, &bytes)))
+        goto err;
```

Comment:
sockerr was last set up at line 126 if we take this error path, which is probably not a useful value anymore.
If we're not going to update it, we should perhaps clear it back to zero closer to when we last set it..

Reply:
It should be updated as the `BIO_write_ex()` could fail.